skipchain/skipchain.go

```diff
@@ -1360,22 +1349,76 @@ func (s *Service) startBFT(proto string, origRoster, newRoster *onet.Roster, msg
 	}
 }
 
-// PropagateSkipBlock will save a new SkipBlock
-func (s *Service) propagateSkipBlock(msg network.Message) {
-	sbs, ok := msg.(*PropagateSkipBlocks)
+// PropagateGenesisHandler will save a new SkipBlock
+func (s *Service) propagateGenesisHandler(msg network.Message) {
+	pg, ok := msg.(*PropagateGenesis)
 	if !ok {
 		log.Error("Couldn't convert to slice of SkipBlocks")
 		return
 	}
-	for _, sb := range sbs.SkipBlocks {
-		if !s.blockIsFriendly(sb) {
-			log.Lvlf2("%s: block is not friendly: %x", s.ServerIdentity(), sb.Hash)
+
+	s.db.Store(pg.Genesis)
+
+	if !pg.Genesis.ParentBlockID.IsNull() {
+		sb := s.db.GetByID(pg.Genesis.ParentBlockID)
+		if sb == nil {
+			log.Lvl3("Unknown parent block. Conode could be out of sync.")
 			return
 		}
+
+		sb.ChildSL = append(sb.ChildSL, pg.Genesis.Hash)
+		s.db.Store(sb)
 	}
-	_, err := s.db.StoreBlocks(sbs.SkipBlocks)
-	if err != nil {
-		log.Error(err)
+}
+
+// PropagateForwardLinkHandler will update the latest block with
+// the new forward link and the new block when given
+func (s *Service) propagateForwardLinkHandler(msg network.Message) {
+	pfl, ok := msg.(*PropagateForwardLink)
+	if !ok {
+		log.Error("Couldn't convert to slice of ForwardLinks")
+		return
+	}
+
+	// Get the block to update the list of FLs
+	sb := s.db.GetByID(pfl.ForwardLink.From)
+	if sb == nil {
```

Comment:
As far as I understand, if `sb == nil`, then we can return here. What usually happens is this (for level >0 link):
1. leader node asks earlier leader with `ForwardSignature` to create a forward signature
2. earlier leader creates forward signature using `forwardLink`
3. all earlier nodes verify the forward-link and _TODO_: store the block
4. the earlier leader sends `propagateForwardLinkHandler` with the forwardlink

So if a node receives a forward-link
- pointing to/from a block it doesn't has, it can ignore it
- it cannot verify, it can ignore it

If you're in doubt, write me on slack, and we'll whiteboard it.